lib/SimpleSAML/Metadata/Signer.php

```diff
@@ -235,35 +235,31 @@ private static function getMetadataSigningAlgorithm(
     public static function sign(string $metadataString, array $entityMetadata, string $type): string
     {
         $config = Configuration::getInstance();
-        $configUtils = new Utils\Config();
+        $cryptoUtils = new Utils\Crypto();
 
         // check if metadata signing is enabled
         if (!self::isMetadataSigningEnabled($config, $entityMetadata, $type)) {
             return $metadataString;
         }
 
         // find the key & certificate which should be used to sign the metadata
-        $keyCertFiles = self::findKeyCert($config, $entityMetadata, $type);
+        $keyCertLocations = self::findKeyCert($config, $entityMetadata, $type);
 
-        $keyFile = $configUtils->getCertPath($keyCertFiles['privatekey']);
-        $fileSystem = new Filesystem();
-        if (!$fileSystem->exists($keyFile)) {
+        $keyLocation = $keyCertLocations['privatekey'];
+        $keyData = $cryptoUtils->retrieveKey($keyLocation);
+        if ($keyData === null) {
             throw new Exception(
-                'Could not find private key file [' . $keyFile . '], which is needed to sign the metadata'
+                'Could not find private key location [' . $keyLocation . '], which is needed to sign the metadata'
```

Comment:
If I understand it correctly, for the default use case the error will no longer show the full filesystem path where it's looking for the certificate, right? I'd prefer that it does because it's a common exception and understanding where to look is very helpful there.

Reply:
The exception text in Metadata/Signer.php will only include the configured location, be it the db or file format, and yes, it won't be fully qualified if the location was configured as relative. However, the error generated in Utils/Crypto.php will contain the fully qualified path:

$location = $configUtils->getCertPath($location);
[...]
Logger::error("failed to read $data_type data from file $location");

So the information will be available. If that's not good enough, I can update retrieveCertOrKey to throw an exception on failure instead of returning null and have the full path bubble up to be caught by Metadata/Signer.php to be used in the text it generates for its exception.

Reply:
Just touching base and seeing if you'd had a chance to consider if the current approach of logging the full path of a missing cert/key in retrieveCertOrKey and just the location in the caller is acceptable or if you want me to update it to raise an exception and have the caller be able to log the full path.

Thanks...

Reply:
Thanks. It's good enough. I'd be fine with merging this if you can resolve the conflicts.

Reply:
I rebased to latest master and it seems fine? Didn't have to manually clean up anything.